### DISCUSSION TITLE
有没有可能将v2ray开发成linux内核驱动呢？

### DISCUSSION THREAD

**Discussion by xuyuji9000** [v2fly/v2ray-core#826] [Ideas]
如果被merge进upstream的话，所有使用linux最新内核版本的机器都可以默认使用v2ray。

可以类比Wireguard, 不过是针对GFW

**Comment by Loyalsoldier** [comment]
有可能，万分之一的可能。

**Reply by xuyuji9000** [reply]
hhh 
因为最近花了很多时间学习linux内核的数据结构，自己也用v2ray，所以会想到这个点，hhh

**Comment by Cwek** [comment]
你至少要得到Linus老大的赞许吧。
Linus就是赞Wireguard写的代码简洁，比大多数的VPN类应用还要精巧才鼓励整合进Linux内核。
而且Linux和Wireguard主要用C编写的，而v2ray是go，还需要做移植工作。
想法很好，但只是想想就好罢了。

**Reply by xuyuji9000** [reply]
先把idea暴露出来，说不定有大神感兴趣呢？

**Reply by xuyuji9000** [reply]
Wireguard也是有go的[user space实现](https://git.zx2c4.com/wireguard-go/about/)的, 不过那是为非Linux用户准备的, Linux的用户被建议直接使用kernel module.

按照这个逻辑的话，v2ray如果被看成一个协议, 那么这个协议是不是也能够通过C重新实现一遍呢？

> 当然这里首先需要回答的问题是，为什么要进入kernel mainline?

> 目前的认知是，进入了kernel mainline的话，v2ray就会随着linux kernel一起被分发到所有使用Linux的机器上，极大的降低了开始使用的门槛，有点类似于进入了app store的ios app，对于终端用户更加友好

**Reply by xuyuji9000** [reply]
Wireguard(wg)的代码确实非常赞, 最近学习Linux的源码，其中很重要的参考就是wg.

**Reply by ygcaicn** [reply]
不可能的，不该在内核的东西。

**Reply by xuyuji9000** [reply]
请问不该在内核的逻辑是？

作为一个内核网络驱动(network driver)也不合理么？

**Comment by okudayukiko** [comment]
Linus的脾氣並不好。而且現在的LKM都有數位簽章機制。

**Comment by Aleksanaa** [comment]
有没有一种可能，并不需要merge进mainline,因为大多数人不需要这个；但确实可以开发成内核模块，应该会有性能提升？
